distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala

```diff
@@ -962,6 +967,61 @@ object Lifecycle extends LifecycleCatsInstances {
     }
   }
 
+  @inline
+  private final def foldMImpl[F[_], A, B](
+    self: Lifecycle[F, A]
+  )(failure: Throwable => Lifecycle[F, B],
+    success: A => Lifecycle[F, B],
+  )(implicit F: QuasiIO[F]
+  ): Lifecycle[F, B] = {
+    import QuasiIO.syntax._
+    new Lifecycle[F, B] {
+      override type InnerResource = AtomicReference[List[() => F[Unit]]]
+
+      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
+        F.bracket(
+          acquire = lifecycle.acquire.flatMap {
+            a =>
+              F.maybeSuspend {
+                // can't use `.updateAndGet` because of Scala.js
+                var oldValue = finalizers.get()
+                while (!finalizers.compareAndSet(oldValue, (() => lifecycle.release(a)) :: oldValue)) {
+                  oldValue = finalizers.get()
+                }
+                a
+              }
+          }
+        )(release = _ => F.unit)(
+          use = use
+        )
+      }
+
+      override def acquire: F[InnerResource] = F.maybeSuspend(new AtomicReference(Nil))
+      override def release(finalizers: InnerResource): F[Unit] = F.suspendF(F.traverse_(finalizers.get())(_.apply()))
+      override def extract[C >: B](finalizers: InnerResource): Either[F[C], C] = Left {
+        F.definitelyRecover {
+          bracketAppendFinalizer(finalizers)(self) {
+            inner1: self.InnerResource =>
+              self.extract(inner1).fold(_.map(success), F pure success(_)).flatMap {
+                that: Lifecycle[F, B] =>
+                  bracketAppendFinalizer(finalizers)(that) {
```

Comment:
@VladPodilnyk It is possible for `failure` and `success` to both run at the same time with this scheme, if the failure occurs at or after this line, which is logically incorrect.